app/com/arpnetworking/metrics/portal/scheduling/impl/JobScheduler.java

```diff
@@ -0,0 +1,225 @@
+                    final Optional<Job<T>> maybeJob = _jobRef.get();
+                    if (!maybeJob.isPresent()) {
+                        LOGGER.error()
+                                .setMessage("no such job")
+                                .addData("jobRef", _jobRef)
+                                .log();
+                        return;
+                    }
+                    final Job<T> job = maybeJob.get();
+                    final Optional<Instant> maybeNextRun = job.getSchedule().nextRun(_jobRef.getLastRun());
+                    if (!maybeNextRun.isPresent()) {
+                        getSelf().tell(PoisonPill.getInstance(), getSelf());
+                        return;
+                    }
+                    final Instant nextRun = maybeNextRun.get();
+                    if (nextRun.isAfter(_clock.instant())) {
+                        // It's not time to execute the job yet, but it might be soon.
+                        // Let's ensure we wake up "exactly" when it's time to execute the job,
+                        //   else we might be up to 1tick late executing it.
+                        //   (Well-- "exactly" is a tall order, but we can at least ensure we wake up
+                        //    _very slightly_ late.)
+                        timeUntilExtraTick(_clock.instant(), nextRun).ifPresent(delta -> {
+                            timers().startSingleTimer("TICK_ONEOFF", Tick.INSTANCE, delta);
+                        });
+                    } else {
+                        final ActorRef sender = getSender();
+                        PatternsCS.pipe(
+                                job.execute(getSelf(), nextRun)
+                                        .handle((result, error) -> {
+                                            if (error == null) {
+                                                _jobRef.jobSucceeded(nextRun, result);
+                                                return new Success<>(_jobRef, nextRun, result);
+                                            } else {
+                                                _jobRef.jobFailed(nextRun, error);
+                                                return new Failure<>(_jobRef, nextRun, error);
+                                            }
+                                        }),
+                                getContext().dispatcher())
+                                .to(sender);
+                        _jobRef.jobStarted(nextRun);
+                    }
+                })
+                .build();
+    }
+
+    protected static final FiniteDuration TICK_INTERVAL = Duration.apply(1, TimeUnit.MINUTES);
+    protected static final FiniteDuration SLEEP_SLOP = Duration.apply(10, TimeUnit.MILLISECONDS);
+    private static final Logger LOGGER = LoggerFactory.getLogger(JobScheduler.class);
+
+    /**
+     * Internal message, telling the scheduler to run any necessary jobs.
+     */
+    protected static final class Tick {
```

Comment:
Is there a reason this isn't private? (or if you need it for tests, package private)